### PR TITLE
1ct best nonce

### PR DESCRIPTION
**Pull request by onemicky**
---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1207517313253443

### CODE REVIEW

File: src/lib/contracts/utils.ts [hunk 1]
```diff
@@ -46,3 +46,63 @@ export async function getGasLimit(
 
   return (gasLimit * 11n) / 10n; // add a 10% buffer
 }
+
+export function getBestNonce(providers: Wallet[]): Promise<number> {
+  return new Promise(async (resolve, reject) => {
+    const results: number[] = [];
+
+    let pending = providers.length;
+    let threshold = 3;
```
[the remark this answers is not in the document]
Comment (gmxer) [reply src/lib/contracts/utils.ts:55]: I think logic should be organized the way that each successful request adds another ~1 seconds to a wait time for other requests. And some total limit (5000ms seems fine). For example let's say If I have two requests:

case 1
```
0000ms r1, r2 started
1000ms r1 fulfilled with 2 (wait for another second for a r2)
1500ms r2 fulfilled with 3 
1500ms return 3
```
case 2
```
0000ms r1, r2 started
1000ms r1 fulfilled with 2 (wait for another second for a r2)
2000ms return 2
2500ms r2 fulfilled with 3 
```
case 3
```
0000ms r1, r2 started
5000ms reject with error
```

[the remark this answers is not in the document]
Comment (gmxer) [reply src/lib/contracts/utils.ts:55]: please also add some tests with 1,2 and 4 wallets

[the remark this answers is not in the document]
Comment (gmxer) [reply src/lib/contracts/utils.ts:55]: also I found a case in current solution that doesn't make sense. It got 2 different nonce but rejected with error:
```
test("weird case", () => {
    const providers: any[] = [
      new MockWallet(1, false, 1),
      new MockWallet(2, true, 4998),
      new MockWallet(3, true, 4999),
    ];
    const res = getBestNonce(providers);
    jest.advanceTimersByTime(5001);
    expect(res).resolves.toBe(2);
  });
```

**Review by gmxer — CHANGES_REQUESTED** [review]

File: src/lib/contracts/utils.ts [hunk 2]
```diff
@@ -46,3 +46,59 @@ export async function getGasLimit(
...
+  const MAX_NONCE_NEEDED = 3;
+  const MAX_WAIT = 5000;
+  const ONE_MORE_WAIT = 1000;
+
+  return new Promise(async (resolve, reject) => {
+    const results: number[] = [];
+    let resolved = false;
+
+    const handleResolve = () => {
+      resolved = true;
```
Comment (gdev8317) [inline src/lib/contracts/utils.ts:60]: a small suggestion, you can clear timeout here, so there will be no need to track `resolved` and check it inside `setResolveTimeout`

File: src/lib/contracts/utils.ts [hunk 3]
```diff
@@ -46,3 +46,63 @@ export async function getGasLimit(
...
+
+    let pending = providers.length;
+    let threshold = 3;
+    let shown = false;
+
+    const handleSuccess = () => {
+      if (!shown) {
+        shown = true;
+        // eslint-disable-next-line no-console
+        console.log("Nonces been received: ", results);
```
[the remark this answers is not in the document]
Comment (onemicky) [reply src/lib/contracts/utils.ts:64]: Sorry for that, it's only for test purposes. Will defenitely remove before merge

File: src/lib/contracts/utils.ts (hunk 2 above)
Comment (gdev8317) [reply src/lib/contracts/utils.ts:60]: discussed it dm, no need to do this

**Review by gmxer — APPROVED** [review]

**Pull request merged**